src/component/componentProps.ts

```diff
@@ -49,24 +42,23 @@ type ExtractCorrectPropType<T> = T extends Function
   ? ExtractFunctionPropType<T>
   : Exclude<T, Function>
 
-// prettier-ignore
 type InferPropType<T> = T extends null
   ? any // null & true would fail to infer
   : T extends { type: null | true }
-    ? any // As TS issue https://github.com/Microsoft/TypeScript/issues/14829 // somehow `ObjectConstructor` when inferred from { (): T } becomes `any` // `BooleanConstructor` when inferred from PropConstructor(with PropMethod) becomes `Boolean`
-    : T extends ObjectConstructor | { type: ObjectConstructor }
-      ? { [key: string]: any }
-      : T extends BooleanConstructor | { type: BooleanConstructor }
-        ? boolean
-        : T extends FunctionConstructor
-          ? Function
-          : T extends Prop<infer V>
-            ? ExtractCorrectPropType<V> : T;
+  ? any // As TS issue https://github.com/Microsoft/TypeScript/issues/14829 // somehow `ObjectConstructor` when inferred from { (): T } becomes `any` // `BooleanConstructor` when inferred from PropConstructor(with PropMethod) becomes `Boolean`
+  : T extends ObjectConstructor | { type: ObjectConstructor }
+  ? Record<string, any>
+  : T extends BooleanConstructor | { type: BooleanConstructor }
+  ? boolean
+  : T extends FunctionConstructor
+  ? Function
+  : T extends Prop<infer V, infer D>
+  ? unknown extends V
+    ? D
+    : ExtractCorrectPropType<V>
+  : T
```

Comment:
Could you keep the format as https://github.com/vuejs/vue-next/blob/c6443a43c9744d8d40f7129a98792debe07a3a82/packages/runtime-core/src/componentProps.ts#L77-L85